pkg/ako-operator/akodeploymentconfig_cluster_mapping_helper.go

```diff
@@ -0,0 +1,174 @@
+// Copyright 2022 VMware, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package ako_operator
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	akoov1alpha1 "github.com/vmware-samples/load-balancer-operator-for-kubernetes/api/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/util/conditions"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// ListAkoDeplymentConfigSelectClusters list all clusters enabled current akodeploymentconfig
+func ListAkoDeplymentConfigSelectClusters(
+	ctx context.Context,
+	kclient client.Client,
+	log logr.Logger,
+	obj *akoov1alpha1.AKODeploymentConfig) (*clusterv1.ClusterList, error) {
+	// get all clusters can be selected by this akodeploymentconfig's cluster selector
+	selector, err := metav1.LabelSelectorAsSelector(&obj.Spec.ClusterSelector)
+	if err != nil {
+		return nil, err
+	}
+	var clusters clusterv1.ClusterList
+	if err := kclient.List(ctx, &clusters, []client.ListOption{
+		client.MatchingLabelsSelector{Selector: selector},
+	}...); err != nil {
+		return nil, err
+	}
+	// remove clusters that:
+	// 1. not ready
+	// 2. management cluster
+	// 3. previously selected by other adc objects
+	var newItems []clusterv1.Cluster
+	var allErrs []error
+	for _, cluster := range clusters.Items {
+		if !SkipCluster(&cluster) {
+			// management cluster can't be selected by other adc objects
+			// except the management cluster AKODeploymentConfig
+			if cluster.Namespace == akoov1alpha1.TKGSystemNamespace &&
+				obj.Name != akoov1alpha1.ManagementClusterAkoDeploymentConfig {
+				continue
+			}
+			adcName, exist := cluster.Labels[akoov1alpha1.AviClusterLabel]
+			// if cluster is already selected by other customized adc objects, skip
+			// only clusters selected by default adc with empty selector object can be overrided
+			if exist && adcName != obj.Name {
+				if !isDefaultADC(adcName) || !defaultADCHasEmptySelector(ctx, kclient) {
```

Comment:
When the default akodeploymentconfig object's selector is not empty, it indicates that this default adc is not the special one that can select all workload clusters, so we treat it as the same as other adc objects.

Here since the cluster is already been selected by another adc, we should not allow it to be selected again.